File: src/filler.py
"""チャンネルが静かになりすぎないように、Tier 2 news の谷間に投稿する軽い息抜き。

- 🐈 The Cat API から猫画像 1 枚
- 📰 Yahoo 話題 / livedoor エンタメ から軽い記事 1 件をランダム紹介

Tier 2 の裏事情解説とは別物なので、LLM 分析は無し。API コストほぼ 0。
"""

from __future__ import annotations

import logging
import os
import random
import sys
from datetime import datetime

import feedparser
import httpx


logger = logging.getLogger(__name__)


CAT_API = "https://api.thecatapi.com/v1/images/search?limit=1&mime_types=jpg,png,gif"

LIGHT_FEEDS: list[tuple[str, str]] = [
    ("Yahoo 話題",          "https://news.yahoo.co.jp/rss/categories/life.xml"),
    ("livedoor エンタメ",   "https://news.livedoor.com/topics/rss/ent.xml"),
    ("Yahoo エンタメ",      "https://news.yahoo.co.jp/rss/topics/entertainment.xml"),
    ("Yahoo IT",            "https://news.yahoo.co.jp/rss/topics/it.xml"),
]

USER_AGENT = "news-bot-filler/0.1 (+https://github.com/klonyapin/news-bot)"


def fetch_cat_image(client: httpx.Client) -> str | None:
    try:
        r = client.get(CAT_API, timeout=10.0)
        r.raise_for_status()
        data = r.json()
        if data and isinstance(data, list):
            return data[0].get("url")
    except Exception as e:
        logger.warning("Cat API failed: %s", e)
    return None


def fetch_light_news(client: httpx.Client) -> tuple[str, str, str] | None:
    """(source_name, title, url) を返す。全 feed 失敗なら None。"""
    feeds = random.sample(LIGHT_FEEDS, len(LIGHT_FEEDS))
    for name, url in feeds:
        try:
            r = client.get(url, headers={"User-Agent": USER_AGENT}, timeout=10.0)
            r.raise_for_status()
            parsed = feedparser.parse(r.text)
            if not parsed.entries:
                continue
            entry = random.choice(parsed.entries[:20])
            title = str(entry.title).strip()
            link = str(entry.link).strip()
            if title and link:
                return name, title, link
        except Exception as e:
            logger.warning("Feed %s failed: %s", name, e)
    return None


def post_to_discord(
    client: httpx.Client,
    webhook: str,
    cat_url: str | None,
    news: tuple[str, str, str] | None,
) -> None:
    now = datetime.now().strftime("%m/%d %H:%M JST")
    content = f"🐈 **今日の休憩**  ·  {now}"
    embed: dict = {"color": 0xFFA07A}

    if cat_url:
        embed["image"] = {"url": cat_url}

    if news:
        src, title, link = news
        embed["title"] = title[:250]
        embed["url"] = link
        embed["footer"] = {"text": src}
        embed["description"] = f"📰 息抜きに 1 本"

    if not cat_url and not news:
        # 両方失敗 = 投稿しない (空メッセージを送らない)
        logger.warning("Both cat and news fetch failed, skipping post")
        return

    payload = {"content": content, "embeds": [embed]}
    r = client.post(webhook, json=payload, timeout=15.0)
    r.raise_for_status()
    logger.info("Posted filler: cat=%s, news=%s", bool(cat_url), news[1][:40] if news else None)


def main() -> None:
    logging.basicConfig(
        level=os.getenv("LOG_LEVEL", "INFO").upper(),
        format="%(asctime)s [%(levelname)s] %(message)s",
        datefmt="%H:%M:%S",
    )
    webhook = os.getenv("DISCORD_WEBHOOK_URL")
    if not webhook:
        logger.error("DISCORD_WEBHOOK_URL not set")
        sys.exit(2)

    with httpx.Client(follow_redirects=True) as client:
        cat_url = fetch_cat_image(client)
        news = fetch_light_news(client)
        post_to_discord(client, webhook, cat_url, news)


if __name__ == "__main__":
    main()

File: worker/src/filler.ts
import { XMLParser } from "fast-xml-parser";

const CAT_API = "https://api.thecatapi.com/v1/images/search?limit=1&mime_types=jpg,png,gif";

const USER_AGENT = "news-bot-quake/0.1 (+https://github.com/klonyapin/news-bot)";

type LightFeed = { name: string; url: string };

const LIGHT_FEEDS: LightFeed[] = [
  { name: "Yahoo 話題",        url: "https://news.yahoo.co.jp/rss/categories/life.xml" },
  { name: "livedoor エンタメ",  url: "https://news.livedoor.com/topics/rss/ent.xml" },
  { name: "Yahoo エンタメ",     url: "https://news.yahoo.co.jp/rss/topics/entertainment.xml" },
  { name: "Yahoo IT",          url: "https://news.yahoo.co.jp/rss/topics/it.xml" },
];

const rssParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  isArray: (name) => name === "item" || name === "entry",
});

type LightNews = { source: string; title: string; link: string };

async function fetchCatImage(): Promise<string | null> {
  try {
    const res = await fetch(CAT_API, { headers: { "User-Agent": USER_AGENT } });
    if (!res.ok) {
      console.warn(`Cat API returned ${res.status}`);
      return null;
    }
    const data = (await res.json()) as Array<{ url?: string }>;
    return data?.[0]?.url ?? null;
  } catch (e) {
    console.warn("Cat API failed:", (e as Error).message);
    return null;
  }
}

async function fetchLightNews(): Promise<LightNews | null> {
  // ランダム順で試して最初に成功した feed から 1 記事ピック
  const shuffled = [...LIGHT_FEEDS].sort(() => Math.random() - 0.5);
  for (const feed of shuffled) {
    try {
      const res = await fetch(feed.url, { headers: { "User-Agent": USER_AGENT } });
      if (!res.ok) continue;
      const xml = await res.text();
      const doc = rssParser.parse(xml);
      const items = (doc?.rss?.channel?.item ?? doc?.["rdf:RDF"]?.item ?? []) as Array<{ title?: string; link?: string }>;
      if (!items.length) continue;
      const top = items.slice(0, 20);
      const pick = top[Math.floor(Math.random() * top.length)];
      const title = String(pick?.title ?? "").trim();
      const link = String(pick?.link ?? "").trim();
      if (title && link) {
        return { source: feed.name, title, link };
      }
    } catch (e) {
      console.warn(`Light feed ${feed.name} failed:`, (e as Error).message);
    }
  }
  return null;
}

function formatJstNow(): string {
  const jst = new Date(Date.now() + 9 * 3600 * 1000);
  const mm = jst.getUTCMonth() + 1;
  const dd = jst.getUTCDate();
  const hh = String(jst.getUTCHours()).padStart(2, "0");
  const min = String(jst.getUTCMinutes()).padStart(2, "0");
  return `${mm}/${dd} ${hh}:${min} JST`;
}

function truncate(s: string, limit: number): string {
  return s.length <= limit ? s : s.slice(0, limit - 1) + "…";
}

export async function postFiller(webhookUrl: string): Promise<void> {
  const [catUrl, news] = await Promise.all([fetchCatImage(), fetchLightNews()]);

  if (!catUrl && !news) {
    console.warn("Both cat and news fetch failed, skipping filler post");
    return;
  }

  const embed: Record<string, unknown> = { color: 0xffa07a };
  if (catUrl) embed.image = { url: catUrl };
  if (news) {
    embed.title = truncate(news.title, 250);
    embed.url = news.link;
    embed.footer = { text: news.source };
    embed.description = "📰 息抜きに 1 本";
  }

  const payload = {
    content: `🐈 **今日の休憩**  ·  ${formatJstNow()}`,
    embeds: [embed],
  };

  const res = await fetch(webhookUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Discord filler post failed ${res.status}: ${body.slice(0, 200)}`);
  }
  console.log(`Filler posted: cat=${!!catUrl}, news=${news?.title.slice(0, 40) ?? "none"}`);
}
